onnxscript/ir/passes/_pass_infra.py

```diff
@@ -0,0 +1,250 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+#
+# This module implements NodeTransformer API described in
+# https://pytorch.org/executorch/stable/compiler-custom-compiler-passes.html
+# for the ONNX IR.
+# The classes {PassResult and PassManager} are derived from
+# https://github.com/pytorch/pytorch/blob/1e47c7b11b312b47a621efd547f5c90081f0d9cb/torch/fx/passes/infra/pass_base.py#L12
+# and
+# https://github.com/pytorch/pytorch/blob/1e47c7b11b312b47a621efd547f5c90081f0d9cb/torch/fx/passes/infra/pass_manager.py#L147
+# The original code is licensed under the PyTorch License https://github.com/pytorch/pytorch/blob/main/LICENSE
+
+"""Passes infrastructure for the IR."""
+
+from __future__ import annotations
+
+import dataclasses
+import logging
+from typing import Callable
+
+__all__ = [
+    "PassBase",
+    "PassResult",
+    "PassManager",
+    "NodeTransformer",
+    # Errors
+    "InvariantError",
+    "PreconditionError",
+    "PostconditionError",
+    "PassError",
+]
+
+import abc
+
+from onnxscript import ir
+
+logger = logging.getLogger(__name__)
+
+
+class InvariantError(Exception):
+    """Raised when an invariant is violated."""
+
+
+class PreconditionError(InvariantError):
+    """Raised when a precondition is violated."""
+
+
+class PostconditionError(InvariantError):
+    """Raised when a postcondition is violated."""
+
+
+class PassError(RuntimeError):
+    """Raised when an error occurs during a pass."""
+
+
+@dataclasses.dataclass
+class PassResult:
+    """Result of a pass.
+
+    Attributes:
+        model: The transformed model.
+        modified: Whether the model was modified.
+    """
+
+    model: ir.Model
+    modified: bool
+
+
+class PassBase(abc.ABC):
+    """Base class for all passes.
+
+    Class attributes:
+        in_place: Whether the pass modifies the model in place.
+    """
+
+    in_place: bool = True
+    modified: bool
+
+    def __call__(self, model: ir.Model) -> PassResult:
+        return self.call(model)
+
+    @abc.abstractmethod
+    def call(self, model: ir.Model) -> PassResult:
+        """The main entry point for the pass."""
+        ...
+
+    def requires(self, model: ir.Model) -> None:
+        """Pre-conditions for the pass.
+
+        This is optional to implement, will be called before call() if run by a pass manager.
+        """
+        del model  # Unused
+
+    def ensures(self, model: ir.Model) -> None:
+        """Post-conditions for the pass.
+
+        This is optional to implement, will be called after call() if run by a pass manager.
+        """
+        del model  # Unused
+
+
+class NodeTransformer(PassBase):
```

Comment:
Trying to understand the relation between this and the existing Visitor class infrastructure. The rest of the infrastructure (PassBase and PassManager) has nothing equivalent in existing code, and looks fine. 

But this is more about how we implement a single pass and it overlaps the existing Optimizer implementation. It would be useful to ensure that we don't end up with multiple ways of doing things.

For example:
* should we have an abstraction that allows users to replace a single node by a list of nodes? (Which simultaneously covers the 1-to-many, 1-to-1, and 1-to-none). Where the infrastructure takes care of revisiting the replacement nodes? 
* Do we need a more general-purpose visitor mechanism? Maybe we can ignore this question, I guess what is being introduced here is a limited visitor abstraction just for nodes.

Reply:
For our discussion: We can improve the class as we migrate the optimizer. I will add a comment cautioning users to not depend on this class yet.